VisualRiders.PointOfSale.Project/Services/DiscountsService.cs

```diff
@@ -0,0 +1,195 @@
+using VisualRiders.PointOfSale.Project.Models;
+using VisualRiders.PointOfSale.Project.Repositories;
+
+namespace VisualRiders.PointOfSale.Project.Services;
+
+public class DiscountsService
+{
+    private readonly DiscountsRepository _repository;
+    private readonly DiscountItemsRepository _discountItemsRepository;
+    private readonly ProductsRepository _productsRepository;
+    private readonly ServicesRepository _servicesRepository;
+    private readonly IMapper _mapper;
+
+    public DiscountsService(
+        DiscountsRepository repository,
+        DiscountItemsRepository discountItemsRepository,
+        ProductsRepository productsRepository,
+        ServicesRepository servicesRepository,
+        IMapper mapper)
+    {
+        _repository = repository;
+        _discountItemsRepository = discountItemsRepository;
+        _productsRepository = productsRepository;
+        _servicesRepository = servicesRepository;
+        _mapper = mapper;
+    }
+
+    public Discount Create(CreateUpdateDiscountDto dto)
+    {
+        var discount = _mapper.Map<Discount>(dto);
+
+        discount.BusinessEntityId = 1;
+        discount.DiscountItems = new List<DiscountItem>();
+        
+        _repository.Add(discount);
+        _repository.SaveChanges();
+
+        return discount;
+    }
+
+    public List<ReadDiscountDto> GetAll()
+    {
+        return _repository.GetAll().Select(_mapper.Map<ReadDiscountDto>).ToList();
+    }
+
+    public ReadDiscountDto? GetById(int id)
+    {
+        return _mapper.Map<ReadDiscountDto>(_repository.GetById(id));
+    }
+
+    public ReadDiscountDto? UpdateById(int id, CreateUpdateDiscountDto dto)
+    {
+        var discount = _repository.GetById(id);
+
+        if (discount == null) return null;
+
+        discount = _mapper.Map<Discount>(dto); 
+
+        _repository.SaveChanges();
+
+        return _mapper.Map<ReadDiscountDto>(discount);
+    }
+
+    public ReadDiscountItemDto? AddItemById(int id, CreateUpdateDiscountItemDto dto)
+    {
+        var discountItem = _mapper.Map<DiscountItem>(dto);
+
+        var discount = _repository.GetById(id);
+
+        if (discount == null) return null;
+
+        discountItem.Discount = discount;
+
+
+        if (dto.ProductId != null)
+        {
+            var product = _productsRepository.GetById(dto.ProductId.Value);
+            if (product == null)
+            {
+                throw new UnprocessableEntity($"Product with Id = {dto.ProductId.Value} does not exist");
+            }
+
+            discountItem.Product = product;
+        }
+
+        if (dto.ServiceId != null)
+        {
+            var service = _servicesRepository.GetById(dto.ServiceId.Value);
+            if (service == null)
+            {
+                throw new UnprocessableEntity($"Product with Id = {dto.ServiceId.Value} does not exist");
+            }
+
+            discountItem.Service = service;
+        }
+
+        if(discount.DiscountItems == null)
```

Comment:
First of all, this is always `null` as you never actually load the list. Secondly, there is no need to fiddle with the `discount` as it is enough to save the `discountItem` and relationship will be properly created.